Created AuthenticationError to be thrown in a case of an Auth fault

EliteScientist opened the pull request:
Authentication Error is thrown whenever any part of the authentication mechanism fails (including requesting the CSRF Token).

Commit:
Created AuthenticationError to be thrown in the case any part of authentication is failing.

Commit:
Fixed linter issues

garethj2 commented on src/errors/AuthenticationError.ts:
```diff
@@ -0,0 +1,11 @@
+export class AuthenticationError extends Error {
+	#cause: Error | undefined
```
It would probably be easier just to make the cause a public property and dispense with the getter.

garethj2 commented on src/client/finCsrfFetch.ts:
```diff
@@ -98,7 +99,7 @@ async function getAttestKey(
 		// If the promise fails then clear the entry from the cache so it can be requested again in future.
 		originAttestKeyPromises.delete(host)
-		throw err
+		throw new AuthenticationError(err as Error)
```
This error is thrown if the attest key can't be acquired. It doesn't necessarily come from an authentication error (HTTP 401 or 403) although it's a likely outcome.

What do you think about changing this to an HttpError that extends Error? This would capture the HTTP status code. A method could be used to assess whether it's an authentication error (i.e. `isAuthenticationError`).

EliteScientist replied:
I agree. this should be an HTTP Error here. Even though it may ultimately result in an AuthenticationError, Authentication Error should be thrown from the authenticator if it detects an auth error.

garethj2 commented on src/errors/AuthenticationError.ts:
```diff
@@ -0,0 +1,11 @@
+export class AuthenticationError extends Error {
```
Copyright required.

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -52,44 +55,71 @@ export async function finAuthFetch(
+			// Response was authenticated, return response
+			return resp
+		} catch (error) {
+			if (error instanceof AuthenticationError) {
```
Following the other comment, this would now check for an `HttpError` and whether it was a real authentication issue via `isAuthenticationError`. 

I would also recommend using a type guard function to ensure it's an `HttpError` instance. Tip: for type guards and JS in general, I try not to use `instanceof` and prefer checking for the existance of properties. I do this just in case there are accidently two versions of `haystack-nclient` loaded into a project (which unfortunately can easily happen in JS).

EliteScientist replied:
I used a global symbol because these types are pretty simple right now. I wanted something a bit concrete. The global symbol will work across versions.

Commit:
Implemented PR Changes:
 - Use type guards
 - Implement HttpError
 - Missing Copyright
 - dispense public getter in favor of readonly public property

Commit:
Rework error handling

EliteScientist commented on src/client/finAuthFetch.ts:
```diff
@@ -2,13 +2,40 @@
+/**
+ * A general authentication error.
+ */
+export class AuthenticationError extends Error {
```
The reason I put this in another file was in the case we needed to throw it in any other part of the system.
